README.md

░█████╗░░█████╗░████████╗░█████╗░████████╗███████╗███╗░░██╗██╗░██████╗
██╔══██╗██╔══██╗╚══██╔══╝██╔══██╗╚══██╔══╝██╔════╝████╗░██║██║██╔════╝
██║░░╚═╝██║░░██║░░░██║░░░███████║░░░██║░░░█████╗░░██╔██╗██║██║╚█████╗░
██║░░██╗██║░░██║░░░██║░░░██╔══██║░░░██║░░░██╔══╝░░██║╚████║██║░╚═══██╗
╚█████╔╝╚█████╔╝░░░██║░░░██║░░██║░░░██║░░░███████╗██║░╚███║██║██████╔╝
░╚════╝░░╚════╝░░░░╚═╝░░░╚═╝░░╚═╝░░░╚═╝░░░╚══════╝╚═╝░░╚══╝╚═╝╚═════╝░


--------------------------------------------------------------------------

# Web crawler

url: [https://www.artwalk.com.br/](https://www.artwalk.com.br/)

# 1. Configuration
Before you run this project and for the proper running of this program you need to set up some variables inside `artwalk/artwalk/settings.py`.
## 1.1 SENTRY
This project utilizes [SENTRY](https://sentry.io/) for error tracking.

- `SENTRY_DSN`
- `SPIDERMON_SENTRY_PROJECT_NAME`
- `SPIDERMON_SENTRY_ENVIRONMENT_TYPE`

## 1.2 GOOGLE CLOUD PLATFORM

- `GCS_PROJECT_ID` 
- `GCP_CREDENTIALS`
- `GCP_STORAGE`
- `GCP_STORAGE_CRAWLER_STATS`
- `IMAGES_STORE`

## 1.3 DISCORD
- `DISCORD_WEBHOOK_URL`
- `DISCORD_THUMBNAIL_URL`
- `SPIDERMON_DISCORD_WEBHOOK_URL`

# 2. Implemented Brands
- artwalk-male-adidas-originals [`ArtWalkAMaleAdidasOriginalsSpider`]
- artwalk-female-adidas-originals [`ArtWalkFemaleAdidasOriginalsSpider`]
- artwalk-kids-adidas-originals [`ArtWalkKidsAdidasOriginalsSpider`]
- artwalk-male-adidas-stansmith [`ArtWalkMaleAdidasStanSmithSpider`]
- artwalk-female-adidas-stansmith [`ArtWalkFemaleAdidasStanSmithSpider`]
- artwalk-kids-adidas-stansmith [`ArtWalkKidsAdidasStanSmithSpider`]
- artwalk-male-adidas-superstar [`ArtWalkMaleAdidasSuperStarSpider`]
- artwalk-female-adidas-superstar [`ArtWalkFemaleAdidasSuperStarSpider`]
- artwalk-kids-adidas-superstar [`ArtWalkKidsAdidasSuperStarSpider`]
- artwalk-adidas-zx [`ArtWalkAdidasZXSpider`]
- artwalk-adidas-ivypark [`ArtWalkAdidasIvyParkSpider`]
- artwalk-female-nike-af1 [`ArtWalkFemaleNikeAF1Spider`]
- artwalk-male-nike-af1 [`ArtWalkMaleNikeAF1Spider`]
- artwalk-male-nike-airmax [`ArtWalkMaleNikeAirMaxSpider`]
- artwalk-female-nike-airmax [`ArtWalkFemaleNikeAirMaxSpider`]
- artwalk-female-nike-jordan [`ArtWalkFemaleNikeJordanSpider`]
- artwalk-male-nike-jordan [`ArtWalkMaleNikeJordanSpider`]
- artwalk-female-nike-lebron [`ArtWalkFemaleNikeLebronSpider`]
- artwalk-male-nike-lebron [`ArtWalkMaleNikeLebronSpider`]
- artwalk-kids-nike-lebron [`ArtWalkKidsNikeLebronSpider`]
- artwalk-female-nike [`ArtWalkFemaleNikeSpider`]
- artwalk-male-nike [`ArtWalkMaleNikeSpider`]
- artwalk-kids-nike [`ArtWalkKidsNikeSpider`]

# 3. Build

```shell
cd artwalk
make docker-build-production
```

# 4. Publish

```shell
make docker-publish-production
```

# 5. Use
The parameter `brand` could receive one of the following values: [`artwalk-male-adidas-originals`, `artwalk-female-adidas-originals`, `artwalk-kids-adidas-originals`, `artwalk-male-adidas-stansmith`, `artwalk-female-adidas-stansmith`, `artwalk-kids-adidas-stansmith`, `artwalk-male-adidas-superstar`, `artwalk-female-adidas-superstar`, `artwalk-kids-adidas-superstar`, `artwalk-adidas-zx`, `artwalk-adidas-ivypark`, `artwalk-female-nike-af1`, `artwalk-male-nike-af1`, `artwalk-male-nike-airmax`, `artwalk-female-nike-airmax`, `artwalk-female-nike-jordan`, `artwalk-male-nike-jordan`, `artwalk-female-nike-lebron`, `artwalk-male-nike-lebron`, `artwalk-kids-nike-lebron`, `artwalk-female-nike`, `artwalk-male-nike`, `artwalk-kids-nike`].

```shell
docker run --shm-size="2g" gcr.io/cotatenis/cotatenis-crawl-artwalk:0.6.0 --brand=artwalk-adidas-zx
```


CHANGELOG.md
# Change Log
Arquivo para documentação das mudanças realizadas ao longo do projeto. O formato desse arquivo é baseado no [Keep a Changelog](http://keepachangelog.com/)
e o presente projeto adota o [Semantic Versioning](http://semver.org/).

## [0.6.0] - 2021-11-17
### [COT-297](https://ecoanalytics.atlassian.net/browse/COT-397)
### Adicionado
- Sobrescrito a função `image_downloaded` do objeto `ImagesPipeline` para garantir a persistência de apenas imagens que ainda não estão salvas no storage.

## [0.5.1] - 2021-11-01
### [COT-294](https://ecoanalytics.atlassian.net/browse/COT-294)
#### Adicionado
- Adicionado o atributo `sku` ao objeto `ArtwalkAdidasItem`.

## [0.5.0] - 2021-10-30
### [COT-294](https://ecoanalytics.atlassian.net/browse/COT-294)
#### Adicionado
- Adicionado as spiders [`ArtWalkAdidasIvyParkSpider`, `ArtWalkFemaleNikeAF1Spider`, `ArtWalkMaleNikeAF1Spider`, `ArtWalkFemaleNikeAirMaxSpider`, `ArtWalkMaleNikeAirMaxSpider`, `ArtWalkFemaleNikeJordanSpider`, `ArtWalkMaleNikeJordanSpider`, `ArtWalkFemaleNikeLebronSpider`, `ArtWalkMaleNikeLebronSpider`, `ArtWalkKidsNikeLebronSpider`, `ArtWalkFemaleNikeSpider`, `ArtWalkMaleNikeSpider`, `ArtWalkKidsNikeSpider`]
#### Alterado
- Alterado lógica de funcionamento do monitor `ItemCountMonitor`.
#### Removido
- Retirado a persistência de imagens no tamanho 800*600.

## [0.4.1] - 2021-10-11
### [COT-236](https://ecoanalytics.atlassian.net/browse/COT-236)
#### Alterado
- Alterado a atribuição da feature `reference_first_image` na coleta.

## [0.4.0] - 2021-10-09
### [COT-202](https://ecoanalytics.atlassian.net/browse/COT-202)
#### Adicionado
- Adicionado a feature `reference_first_image` ao objeto `ArtwalkAdidasItem`. 
#### Alterado
- Alterado a configuração `IMAGES_THUMBS` para salvar imagens no padrão 400x400.

## [0.3.1] - 2021-10-09
### [COT-197](https://ecoanalytics.atlassian.net/browse/COT-149)
#### Alterado
- Alterado o parâmetro `SPIDERMON_MIN_ITEMS` para 300 da `ArtWalkAMaleAdidasOriginalsSpider`.

## [0.3.0] - 2021-10-04
### [COT-149](https://ecoanalytics.atlassian.net/browse/COT-149)
#### Adicionado
- Adicionado monitoramento por [spidermon](https://github.com/scrapinghub/spidermon).
- Adicionado monitoramento por [Sentry](https://sentry.io).

## [0.2.0] - 2021-09-23
### [COT-114](https://ecoanalytics.atlassian.net/browse/COT-114)
#### Adicionado
- Adicionado a classe `ArtWalkImagePipeline` para realizar coleta das imagens dos produtos.
- Adicionado a dependência `Pillow==8.3.2`.
- Adicionado os campos `image_urls` e `image_uris` ao objeto `ArtwalkAdidasItem`.
#### Removido
- Removido a dependência `scrapy-fieldstats`.




artwalk/artwalk/settings.py
from shutil import which
BOT_NAME = 'artwalk'
VERSION = "0-6-0"
SPIDER_MODULES = ['artwalk.spiders']
NEWSPIDER_MODULE = 'artwalk.spiders'
USER_AGENT = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/92.0.4515.131 Safari/537.36'

# Obey robots.txt rules
ROBOTSTXT_OBEY = False
SELENIUM_DRIVER_NAME = 'chrome'
SELENIUM_DRIVER_EXECUTABLE_PATH = which('chromedriver')
SELENIUM_DRIVER_ARGUMENTS=['--headless']  # '--headless' if using chrome instead of firefox
DOWNLOADER_MIDDLEWARES = {
    'scrapy_selenium.SeleniumMiddleware': 800
}
MAGIC_FIELDS = {
    "timestamp": "$isotime",
    "spider": "$spider:name",
    "url": "$response:url",
}
SPIDER_MIDDLEWARES = {
    "scrapy_magicfields.MagicFieldsMiddleware": 100,
}
SPIDERMON_ENABLED = True
EXTENSIONS = {
    'artwalk.extensions.SentryLogging' : -1,
    'spidermon.contrib.scrapy.extensions.Spidermon': 500,
}
ITEM_PIPELINES = {
    "artwalk.pipelines.DiscordMessenger" : 100,
    "artwalk.pipelines.ArtWalkImagePipeline" : 200,
    "artwalk.pipelines.GCSPipeline": 300,
}
SPIDERMON_SPIDER_CLOSE_MONITORS = (
'artwalk.monitors.SpiderCloseMonitorSuite',
)

SPIDERMON_VALIDATION_DROP_ITEMS_WITH_ERRORS = False
SPIDERMON_PERIODIC_MONITORS = {
'artwalk.monitors.PeriodicMonitorSuite': 30, # time in seconds
}
SPIDERMON_CUSTOM_MIN_ITEMS = {
    'artwalk-male-adidas-originals' : 300,
    'artwalk-adidas-ivypark' : 1,
    'artwalk-female-adidas-originals' : 200,
    'artwalk-kids-adidas-originals' : 5,
    'artwalk-female-adidas-stansmith' : 7,
    'artwalk-kids-adidas-stansmith' : 1,
    'artwalk-female-adidas-stansmith' : 10,
    'artwalk-male-adidas-stansmith' : 5,
    'artwalk-female-adidas-superstar' : 200,
    'artwalk-kids-adidas-superstar' : 2,
    'artwalk-male-adidas-superstar' : 40,
    'artwalk-adidas-zx' : 40,
    'artwalk-male-nike-jordan' : 130,
    'artwalk-female-nike-jordan' : 10,
    'artwalk-male-nike-af1' : 90,
    'artwalk-female-nike-af1' : 40,
    'artwalk-male-nike-airmax' : 5,
    'artwalk-female-nike-airmax' : 80,
    'artwalk-male-nike-lebron' : 45,
    'artwalk-female-nike-lebron' : 10,
    'artwalk-kids-nike-lebron' : 4,
    'artwalk-male-nike' : 750,
    'artwalk-female-nike' : 250,
    'artwalk-kids-nike' : 15,
}
SENTRY_DSN = ""
SPIDERMON_SENTRY_PROJECT_NAME = ""
SPIDERMON_SENTRY_ENVIRONMENT_TYPE = ""
#THROTTLE
AUTOTHROTTLE_ENABLED = True
AUTOTHROTTLE_START_DELAY = 2
AUTOTHROTTLE_MAX_DELAY = 5

#GCP
GCS_PROJECT_ID = ""
GCP_CREDENTIALS = ""
GCP_STORAGE = ""
GCP_STORAGE_CRAWLER_STATS = ""
#FOR IMAGE UPLOAD
IMAGES_STORE = ""
IMAGES_THUMBS = {
    '400_400': (400, 400),
}
#DISCORD
DISCORD_WEBHOOK_URL = ""
DISCORD_THUMBNAIL_URL = ""
SPIDERMON_DISCORD_WEBHOOK_URL = ""

#LOGGING
LOG_LEVEL = 'INFO'